zcash_primitives/src/transaction/builder.rs

```diff
@@ -504,17 +520,27 @@ impl<'a, P: consensus::Parameters, R: RngCore + CryptoRng> Builder<'a, P, R> {
         let transparent_bundle = self.transparent_builder.build();
 
         let mut rng = self.rng;
-        let mut ctx = prover.new_sapling_proving_context();
-        let sapling_bundle = self
+        let (sapling_bundle, tx_metadata) = match self
             .sapling_builder
-            .build(
-                prover,
-                &mut ctx,
-                &mut rng,
-                self.target_height,
-                self.progress_notifier.as_ref(),
-            )
-            .map_err(Error::SaplingBuild)?;
+            .build::<SP, OP, _>(&mut rng, self.target_height)
+            .map_err(Error::SaplingBuild)?
+            .map(|(bundle, tx_metadata)| {
+                // We need to create proofs before signatures, because we still support
+                // creating V4 transactions, which commit to the Sapling proofs in the
+                // transaction digest.
```

Comment:
If you look a few lines lower, you'll see:
```rust
b.create_proof(&orchard::circuit::ProvingKey::build(), &mut rng)
```
which is where the Orchard parameters are being constructed on-the-fly like I mentioned above.

Reply:
Separately, yes I think we should at some point deprecate creating v4 transactions. But that will require additional refactoring work elsewhere, because currently the type system forces us to always build proofs before constructing `TransactionData`, so that we are able to derive the txid no matter what version is used. We will likely need to introduce a type-level distinction between pre-v5 and v5+ transactions to address this (instead of always resolving versions at runtime like we currently do).